### ISSUE TITLE
Context.typeCheck fails for macro annotations

### ISSUE THREAD

**Issue by rcano** [scalamacros/paradise#1]
As you can see in this gist, with the most basic annotation macro it throws FatalError

```
import language.experimental.macros
import scala.reflect.macros.Context

class findInnerClasses extends annotation.StaticAnnotation {
  def macroTransform(annottees: Any*) = macro Macros.findInnerClasses
}

object Macros {
  def findInnerClasses(c: Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
    import c.universe._
    val anons = annottees.flatMap{e => 
      e.tree.collect {
        case b@q"""new $tpe[..$tparams](...$vparamss) { ..$body }""" => tpe -> body
        case a@q"""new $tpe(...$vparamss) { ..$body }""" => tpe -> body
    }}.filterNot(_ == null)
    println("Anons found: ")
    for ((tpe, body) <- anons) {
      println("Found: " + tpe)
      for (statement <- body) {
        println("Trying to tc: " + statement)
        if (c.typeCheck(statement).isEmpty) {
          //try to find the broken piece
          statement foreach (t => t foreach (st => println(st + " ---- " + st.symbol)))
        }
      }
    }
    annottees.head
  }
}
```

```
@findInnerClasses object Test extends App {
  new java.util.ArrayList(33)
  new java.util.ArrayList[String](20) {
    def doTheBarrelRoll() {
      println("Do the barrel rol")
      broken
    }
  }
  new javax.swing.JLabel("Dont ever do this, seriouslyss") {
    setText("Why are you still doing this???")
    def multiDef {
      println("the")
      println("Multidef")
    }
  }
}
```

```
18:36 ~/Projects/Paradise/sandbox/src/main/scala (2.10.x)$ ss
Anons found:
Found: java.util.ArrayList
Trying to tc: def doTheBarrelRoll(): scala.Unit = {
  println("Do the barrel rol");
  broken
}
error: no-symbol does not have an owner
error: exception during macro expansion:
scala.reflect.internal.FatalError:
     while compiling: Test.scala
        during phase: typer
     library version: version 2.10.3-20130808-081556-ed5c1abbfc
    compiler version: version 2.10.3-20130808-081556-ed5c1abbfc
  reconstructed args: -Xplugin:/Users/xeno_by/Projects/Paradise/plugin/target/scala-2.10/macro-paradise_2.10.2-2.0.0-SNAPSHOT.jar -language:experimental.macros

  last tree to typer: Ident(broken)
              symbol: value <error> in class <error> (flags: <synthetic> <is_error>)
   symbol definition: val <error>: <error>
                 tpe: <error>
       symbol owners: value <error> -> class <error> -> package <root>
      context owners:

== Enclosing template or block ==

Block(
  Apply(
    "println"
    "Do the barrel rol"
  )
  "broken" // val <error>: <error> in class <error>, tree.tpe=<error>
)

== Expanded type of tree ==

<error>

no-symbol does not have an owner
    at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:49)
    at scala.tools.nsc.Global.abort(Global.scala:254)
    at scala.reflect.internal.Symbols$NoSymbol.owner(Symbols.scala:3248)
    at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2271)
    at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5535)
    at org.scalalang.macroparadise.typechecker.Analyzer$$anon$1.org$scalalang$macroparadise$typechecker$Typers$ParadiseTyper$$super$typed1(Analyzer.scala:17)
    at org.scalalang.macroparadise.typechecker.Typers$ParadiseTyper$class.typed1(Typers.scala:44)
    at org.scalalang.macroparadise.typechecker.Analyzer$$anon$1.typed1(Analyzer.scala:17)
    at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5608)
    at scala.reflect.macros.runtime.Typers$$anonfun$typeCheck$2$$anonfun$5.apply(Typers.scala:25)
    at scala.reflect.macros.runtime.Typers$$anonfun$typeCheck$2$$anonfun$5.apply(Typers.scala:25)
    at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:727)
    at scala.reflect.macros.runtime.Typers$$anonfun$typeCheck$2.apply(Typers.scala:25)
    at scala.reflect.macros.runtime.Typers$$anonfun$typeCheck$2.apply(Typers.scala:25)
    at scala.tools.nsc.typechecker.Contexts$Context.withMacrosEnabled(Contexts.scala:242)
    at scala.reflect.macros.runtime.Typers$$anonfun$3.apply(Typers.scala:18)
    at scala.reflect.macros.runtime.Typers$$anonfun$3.apply(Typers.scala:18)
    at scala.reflect.macros.runtime.Typers$$anonfun$wrapper$1$1.apply(Typers.scala:19)
    at scala.reflect.macros.runtime.Typers$$anonfun$wrapper$1$1.apply(Typers.scala:19)
    at scala.tools.nsc.typechecker.Contexts$Context.withImplicitsEnabled(Contexts.scala:211)
    at scala.reflect.macros.runtime.Typers$$anonfun$1.apply(Typers.scala:17)
    at scala.reflect.macros.runtime.Typers$$anonfun$1.apply(Typers.scala:17)
    at scala.reflect.macros.runtime.Typers$class.wrapper$1(Typers.scala:19)
    at scala.reflect.macros.runtime.Typers$class.typeCheck(Typers.scala:25)
    at scala.reflect.macros.runtime.Context.typeCheck(Context.scala:6)
    at scala.reflect.macros.runtime.Context.typeCheck(Context.scala:6)
    at Macros$$anonfun$findInnerClasses$2$$anonfun$apply$2.apply(Macros.scala:36)
    at Macros$$anonfun$findInnerClasses$2$$anonfun$apply$2.apply(Macros.scala:34)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at Macros$$anonfun$findInnerClasses$2.apply(Macros.scala:34)
    at Macros$$anonfun$findInnerClasses$2.apply(Macros.scala:32)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at Macros$.findInnerClasses(Macros.scala:32)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at scala.tools.nsc.typechecker.Macros$$anonfun$scala$tools$nsc$typechecker$Macros$$macroRuntime$3$$anonfun$apply$8.apply(Macros.scala:544)
    at scala.tools.nsc.typechecker.Macros$$anonfun$scala$tools$nsc$typechecker$Macros$$macroRuntime$3$$anonfun$apply$8.apply(Macros.scala:544)
    at scala.tools.nsc.typechecker.Macros$class.scala$tools$nsc$typechecker$Macros$$macroExpandWithRuntime(Macros.scala:828)

Test.scala:6: error: not found: value broken
      broken
      ^
three errors found
```

**Comment by xeno-by** [comment]
That's a limitation of the way how typechecking is implemented in the compiler.

It'd be hard to fix this in a compiler plugin, but luckily there's a simple workaround: you just need to wrap the expression you want to typecheck in a `Block`, and everything will work out.

Two more things that you might to consider:
1) By default `c.typeCheck` crashes on a typecheck error. You might consider passing `silent = true` in order to prevent crashes and get `EmptyTree`s instead.
2) Most often, typechecking mutates its argument in place. In order to avoid that, call `duplicate` on a tree before passing it to `typeCheck`.

**Comment by rcano** [comment]
Oh, thanks for the info, still, if it is that easy to work around, it sounds like something the compiler should do by itself. Also, in this gist I didn't put it, but  simply trying to typeCheck the whole annottee like:

c.typeCheck(annottees.head.tree)

would also throw a similar exception, and this time, the compiler was receiving a ClassDef, is that also fixed by using a block?

**Comment by xeno-by** [comment]
Yes.

**Comment by Sciss** [comment]
Sorry, I don't understand. Should the following work?

```
c.typeCheck(Block(annottees.head.tree))
```

Because that just gives me `Unit`, instead of the class. If I use

```
c.typeCheck(Block(Nil, v))
```

That gives `AssertionError` again.

**Comment by xeno-by** [comment]
Could you post the input and the output trees printed using showRaw?

**Comment by Sciss** [comment]
So I have

```
import scala.annotation.StaticAnnotation
import scala.reflect.macros._
import language.experimental.macros

class mkCompanion extends StaticAnnotation {
  def macroTransform(annottees: Any*) = macro mkCompanionMacro.impl
}
object mkCompanionMacro {
  def impl(c: Context)(annottees: c.Expr[Any]*): c.Expr[Any] = {
    import c.universe._
    val inputs = annottees.map(_.tree).toList
    val (validAnnottees, _) = inputs partition {
      case x: ClassDef => true
      case _ => false
    }
    for (v <- validAnnottees ) {
      val tpe = v.tpe // <- tpe is null as the annotated type is not yet type checked!
      val tpe2 = if (tpe == null) {
          println(s"---- Input ----\n${showRaw(v)}")
          val blk = c.typeCheck(Block(v))
          val Block((cd1 @ ClassDef(_, _, _, _)) :: Nil, _) = blk
          println(s"---- Output ----\n${showRaw(blk)}")
          cd1.tpe // <- fails with a compiler error (assertion failure)
        }
        else
          tpe
      println(s"type = '$tpe2'")
    }
    c.Expr[Any](Literal(Constant()))
  }
}
```

The `Unit` just came from `Block`. But still, `cd1.tpe` is not resolved:

```
@mkCompanion class Baz
```

Output:

```
---- Input ----
ClassDef(Modifiers(), TypeName("Baz"), List(), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List(DefDef(Modifiers(), nme.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(pendingSuperCall), Literal(Constant(())))))))
---- Output ----
Block(List(ClassDef(Modifiers(), de.sciss.lucre.macros.Baz, List(), Template(List(Select(Ident(scala), TypeName("AnyRef"))), noSelfType, List(DefDef(Modifiers(), nme.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(Apply(Select(Super(This(TypeName("Baz")), tpnme.EMPTY), nme.CONSTRUCTOR), List())), Literal(Constant(())))))))), Literal(Constant(())))
type = '<notype>'
```

This is with Scala 2.11.0-M7 and Macro Paradise 2.0.0-SNAPSHOT

**Comment by xeno-by** [comment]
I'll answer at SO: http://stackoverflow.com/questions/21044957/type-of-a-macro-annottee

**Comment by xeno-by** [comment]
https://github.com/scala/scala/pull/3513

**Comment by xeno-by** [comment]
https://github.com/scala/scala/pull/3551

**Comment by xeno-by** [comment]
Since this is fixed in 2.11.0-RC1 and is about to be merged into 2.10.5, I'm closing this issue.
